### ISSUE TITLE
Install new software feature loop indefinitely in site search

### ISSUE THREAD

**Issue by Graou74** [eclipse-equinox/p2#307]
The problem occurs on version 4.28.0.

To reproduce:

Open the "Available software" window via the "Help/Install New Software ..." menu.
Select "--All Available Sites--" from the "Work With:" combobox or any other line, the behavior remains the same.
The animation for waiting shows that the search is in progress without ever ending (no list of results or a message indicating a problem during the search).
I tested on version 4.28.0 which I have been using for several weeks then on the same downloaded today and the problem remains the same.
On Eclipse version 4.26.0 there is no such problem.

I'm on macOS version 12.6.7 and the JRE is version 19.0.2.

**Comment by merks** [comment]
That could be a humongous list that could (normally) take a very long time to load!  

Could you please also open the Progress View and show a screen capture of what's gone on there?  It shows which site is being contacted like this for me:

![image](https://github.com/eclipse-equinox/p2/assets/208716/7fe192f8-59a9-4d74-93c8-28f4db86e291)

For me too, it takes really long, but eventually finishes.  I'm not sure in your case how long is you "all sites" list and I'm not sure which of those sites in that list are slow.  You should be able to see progress or you should see that it's stuck "forever" on some site and which site that is.

**Comment by Graou74** [comment]
Here is a screenshot in both versions of Eclipse, the one that doesn't seem to work and the one that does.
I imagine that if I choose only one plugin (here "AgileJ - http://www.agilej.com/structureViews") the search time should be acceptable as it is with version 4.26.0 (which by the way also runs in a reasonable time when the search is on the list of all sites). But with version 4.28.0, after at least an hour, for one site only, I suspect that a lack of result is a bug.

![Capture d’écran 2023-08-21 à 16 56 47](https://github.com/eclipse-equinox/p2/assets/54896431/8299849c-2cc9-42ac-aea8-abc49ce74c1e)

**Comment by merks** [comment]
Thank you for the details!

One things that's changed in 4.28 is that http sites are forced to use https (for your safety and comfort):

https://eclipse.dev/eclipse/news/4.28/platform.php#force-https

I can easily reproduce the problem that the site never seems to load using just the Repository Explorer:

![image](https://github.com/eclipse-equinox/p2/assets/208716/f8d7676a-b73a-4a86-ba4f-0ce2270dce8b)

So something is going very wrong with this.  In the worst case, it should timeout (two minutes I think), but it doesn't even seem to do that.  I'll leave it running and see if it ever does something.  

Oh wait, this just showed up in the log:
```
org.apache.hc.client5.http.HttpHostConnectException: Connect to https://www.agilej.com:443 [www.agilej.com/52.95.191.64, www.agilej.com/52.95.191.20, www.agilej.com/52.95.149.143, www.agilej.com/52.95.149.11, www.agilej.com/52.95.149.103, www.agilej.com/52.95.142.36, www.agilej.com/52.95.143.20, www.agilej.com/52.95.191.0] failed: Connection timed out: no further information
	at java.base/sun.nio.ch.Net.pollConnect(Native Method)
	at java.base/sun.nio.ch.Net.pollConnectNow(Net.java:672)
	at java.base/sun.nio.ch.NioSocketImpl.timedFinishConnect(NioSocketImpl.java:549)
	at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:597)
	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327)
	at java.base/java.net.Socket.connect(Socket.java:633)
	at org.apache.hc.client5.http.ssl.SSLConnectionSocketFactory$1.run(SSLConnectionSocketFactory.java:219)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:569)
	at org.apache.hc.client5.http.ssl.SSLConnectionSocketFactory.connectSocket(SSLConnectionSocketFactory.java:216)
	at org.apache.hc.client5.http.impl.io.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:148)
	at org.apache.hc.client5.http.impl.io.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:396)
	at org.apache.hc.client5.http.impl.classic.InternalExecRuntime.connectEndpoint(InternalExecRuntime.java:158)
	at org.apache.hc.client5.http.impl.classic.InternalExecRuntime.connectEndpoint(InternalExecRuntime.java:168)
	at org.apache.hc.client5.http.impl.classic.ConnectExec.execute(ConnectExec.java:136)
	at org.apache.hc.client5.http.impl.classic.ExecChainElement.execute(ExecChainElement.java:51)
	at org.apache.hc.client5.http.impl.classic.ExecChainElement$1.proceed(ExecChainElement.java:57)
	at org.apache.hc.client5.http.impl.classic.ProtocolExec.execute(ProtocolExec.java:175)
	at org.apache.hc.client5.http.impl.classic.ExecChainElement.execute(ExecChainElement.java:51)
	at org.apache.hc.client5.http.impl.classic.ExecChainElement$1.proceed(ExecChainElement.java:57)
	at org.apache.hc.client5.http.impl.classic.HttpRequestRetryExec.execute(HttpRequestRetryExec.java:96)
	at org.apache.hc.client5.http.impl.classic.ExecChainElement.execute(ExecChainElement.java:51)
	at org.apache.hc.client5.http.impl.classic.ExecChainElement$1.proceed(ExecChainElement.java:57)
	at org.apache.hc.client5.http.impl.classic.ContentCompressionExec.execute(ContentCompressionExec.java:133)
	at org.apache.hc.client5.http.impl.classic.ExecChainElement.execute(ExecChainElement.java:51)
	at org.apache.hc.client5.http.impl.classic.ExecChainElement$1.proceed(ExecChainElement.java:57)
	at org.apache.hc.client5.http.impl.classic.RedirectExec.execute(RedirectExec.java:115)
	at org.apache.hc.client5.http.impl.classic.ExecChainElement.execute(ExecChainElement.java:51)
	at org.apache.hc.client5.http.impl.classic.InternalHttpClient.doExecute(InternalHttpClient.java:170)
	at org.apache.hc.client5.http.impl.classic.CloseableHttpClient.execute(CloseableHttpClient.java:75)
	at org.eclipse.ecf.provider.filetransfer.httpclient5.HttpClientRetrieveFileTransfer.performConnect(HttpClientRetrieveFileTransfer.java:1009)
	at org.eclipse.ecf.provider.filetransfer.httpclient5.HttpClientRetrieveFileTransfer$1.performFileTransfer(HttpClientRetrieveFileTransfer.java:997)
	at org.eclipse.ecf.filetransfer.FileTransferJob.run(FileTransferJob.java:76)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
```
But I think there are 3 retries...  So let's see later.

**Comment by Graou74** [comment]
Alright, everything is explained now.
I will contact AgileJ to let them know.

**Comment by Graou74** [comment]
This is for the case where the address used is https:// (https://www.agilej.com/structureViews) but this address does not exist.
If the address at http:// (http://www.agilej.com/structureViews) exists, I have the impression that the loop is infinite.

**Comment by merks** [comment]
I try again now with http:

![image](https://github.com/eclipse-equinox/p2/assets/208716/ab1428cf-7ceb-43cc-8fdf-5ec303c17fa7)

Under the covers, that will use https instead, but maybe the timeout is handled poorly.

Which product exactly are you using?

**Comment by merks** [comment]
Already I have my first timeout:

![image](https://github.com/eclipse-equinox/p2/assets/208716/25c1826e-fcba-41c1-a7b7-dfa374781aec)

**Comment by merks** [comment]
I think it gave up trying to get the p2.index and is trying the jar now:

![image](https://github.com/eclipse-equinox/p2/assets/208716/141b6a10-8434-43bb-8269-cd1a249189bb)

**Comment by merks** [comment]
It tries different file types and each one must fail (timeout):

![image](https://github.com/eclipse-equinox/p2/assets/208716/37d76b14-cf67-43a7-9f3e-b656d9e6dde0)

**Comment by merks** [comment]
Finally failure:

![image](https://github.com/eclipse-equinox/p2/assets/208716/8ac4d210-c942-47db-a544-47d99d734ca4)

That took long, but it terminated...

Of course this site will never work and will always be slow to fail, depending on the timeout value.

**Comment by Graou74** [comment]
Indeed, it is not an infinite loop.
I left the search running last night and this morning I got an error message:

![Capture d’écran 2023-08-22 à 08 59 47](https://github.com/eclipse-equinox/p2/assets/54896431/256377cd-1861-47f6-97d6-40bb0320254b)

I reported the change of situation to AgileJ, there was an email address on the site.

However, I think it's not a good idea to automatically convert http:// to https:// because other sites will probably be in the same situation, with all the hazards that entails.
Instead, I would see a warning that http:// addresses are simply no longer allowed and are being ignored. The user can if he wants to change http:// to https:// and know that the address potentially does not exist and understand that the search is very long in this case.

**Comment by merks** [comment]
Sorry, the https changes are related to CVEs, so we can no longer leave the vulnerability exposed by default.  Yes, other sites were affected, but other sites failed quickly, not so slowly like this one,  and those sites have been fixed because people complained to the site owners.

Also, 4.28 was already released, so changing course now for 4.29 doesn't actually make sense.

I'm very sorry you were bitten by this.   😢 

You can change your preferences to use http if you're comfortable with using http as described here:

https://eclipse.dev/eclipse/news/4.28/platform.php#force-https

**Comment by Graou74** [comment]
Hello Mr Merks,

Don't be sorry, as you may have noticed, although an avid Eclipse user, I still have a lot to learn about this wonderful product.

I indeed spent part of my day yesterday and this morning on what I believed to be a bug and which finally turns out to be a lack of knowledge.

In the end, I was not bitten by this situation and what emerges is that I once again see the tremendous efficiency and responsiveness of Eclipse's development.

So ... thank you.

Best regards.

Frank Beghini

----- Mail original -----
De: "Ed Merks" ***@***.***>
À: "eclipse-equinox/p2" ***@***.***>
Cc: "Graou74" ***@***.***>, "Author" ***@***.***>
Envoyé: Mardi 22 Août 2023 09:36:05
Objet: Re: [eclipse-equinox/p2] Install new software feature loop indefinitely in site search (Issue #307)





Sorry, the https changes are related to CVEs, so we can no longer leave the vulnerability exposed by default. Yes, other sites were affected, but other sites failed quickly, not so slowly like this one, and those sites have been fixed because people complained to the site owners. 

Also, 4.28 was already released, so changing course now for 4.29 doesn't actually make sense. 

I'm very sorry you were bitten by this. 😢 

You can change your preferences to use http if you're comfortable with using http as described here: 

https://eclipse.dev/eclipse/news/4.28/platform.php#force-https
